From: Bogdan Mircea <mirceapetrebogdan@gmail.com>
Subject: [PATCH] hostapd: check VHT capabilities in host byte order

OpenWrt's inherited extended-NSS patch checks host VHT capability words.
Those words and their masks are in host byte order. Converting a word
before masking confuses short-GI-160 with extended NSS on big-endian hosts
and incorrectly changes the advertised width and operation fields.

Upstream-Status: Pending
Signed-off-by: Bogdan Mircea <mirceapetrebogdan@gmail.com>
---

--- a/src/ap/ieee802_11_ht.c
+++ b/src/ap/ieee802_11_ht.c
@@ -127,7 +127,7 @@
 u8 * hostapd_eid_ht_operation(struct hostapd_data *hapd, u8 *eid)
 {
 	struct ieee80211_ht_operation *oper;
-	le32 vht_capabilities_info;
+	u32 vht_capabilities_info;
 	u8 *pos = eid;
 	u8 chwidth;
 
@@ -145,7 +145,7 @@
 	oper->operation_mode = host_to_le16(hapd->iface->ht_op_mode);
 	set_ht_param(hapd, oper);
 
-	vht_capabilities_info = host_to_le32(hapd->iface->current_mode->vht_capab);
+	vht_capabilities_info = hapd->iface->current_mode->vht_capab;
 	chwidth = hostapd_get_oper_chwidth(hapd->iconf);
 	if (vht_capabilities_info & VHT_CAP_EXTENDED_NSS_BW_SUPPORT
 		&& ((chwidth == CHANWIDTH_160MHZ) || (chwidth == CHANWIDTH_80P80MHZ))) {
--- a/src/ap/ieee802_11_vht.c
+++ b/src/ap/ieee802_11_vht.c
@@ -72,14 +72,14 @@
 	}
 
 	chwidth = hostapd_get_oper_chwidth(hapd->iconf);
-	if (((host_to_le32(mode->vht_capab)) & VHT_CAP_EXTENDED_NSS_BW_SUPPORT)
+	if ((mode->vht_capab & VHT_CAP_EXTENDED_NSS_BW_SUPPORT)
 		&& ((chwidth == CHANWIDTH_160MHZ) || (chwidth == CHANWIDTH_80P80MHZ))) {
-		cap->vht_capabilities_info |= VHT_CAP_EXTENDED_NSS_BW_SUPPORT;
+		cap->vht_capabilities_info |= host_to_le32(VHT_CAP_EXTENDED_NSS_BW_SUPPORT);
 		cap->vht_capabilities_info &= ~(host_to_le32(VHT_CAP_SUPP_CHAN_WIDTH_160_80PLUS80MHZ));
 		cap->vht_capabilities_info &= ~(host_to_le32(VHT_CAP_SUPP_CHAN_WIDTH_160MHZ));
 		cap->vht_capabilities_info &= ~(host_to_le32(VHT_CAP_SUPP_CHAN_WIDTH_MASK));
 	} else {
-		cap->vht_capabilities_info &= ~VHT_CAP_EXTENDED_NSS_BW_SUPPORT_MASK;
+		cap->vht_capabilities_info &= ~host_to_le32(VHT_CAP_EXTENDED_NSS_BW_SUPPORT_MASK);
 	}
 
 	/* Supported MCS set comes from hw */
@@ -94,7 +94,7 @@
 u8 * hostapd_eid_vht_operation(struct hostapd_data *hapd, u8 *eid)
 {
 	struct ieee80211_vht_operation *oper;
-	le32 vht_capabilities_info;
+	u32 vht_capabilities_info;
 	u8 *pos = eid;
 	enum oper_chan_width oper_chwidth =
 		hostapd_get_oper_chwidth(hapd->iconf);
@@ -133,7 +133,7 @@
 	oper->vht_op_info_chan_center_freq_seg1_idx = seg1;
 
 	oper->vht_op_info_chwidth = oper_chwidth;
-	vht_capabilities_info = host_to_le32(hapd->iface->current_mode->vht_capab);
+	vht_capabilities_info = hapd->iface->current_mode->vht_capab;
 	if (oper_chwidth == CONF_OPER_CHWIDTH_160MHZ) {
 		/*
 		 * Convert 160 MHz channel width to new style as interop
